src/query/storage/promremote/storage.go

```diff
@@ -0,0 +1,215 @@
+// Copyright (c) 2021  Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package promremote
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/uber-go/tally"
+
+	"github.com/m3db/m3/src/query/block"
+	"github.com/m3db/m3/src/query/storage"
+	"github.com/m3db/m3/src/query/storage/m3/consolidators"
+	"github.com/m3db/m3/src/query/storage/m3/storagemetadata"
+	xerrors "github.com/m3db/m3/src/x/errors"
+	"github.com/m3db/m3/src/x/instrument"
+	xhttp "github.com/m3db/m3/src/x/net/http"
+)
+
+const metricsScope = "prom_remote_storage"
+
+var errorReadingBody = []byte("error reading body")
+
+// NewStorage returns new Prometheus remote write compatible storage
+func NewStorage(opts Options) (storage.Storage, error) {
+	client := xhttp.NewHTTPClient(opts.httpOptions)
+	scope := opts.scope.SubScope(metricsScope)
+	s := &promStorage{opts: opts, client: client, endpointMetrics: initEndpointMetrics(opts.endpoints, scope)}
+	return s, nil
+}
+
+type promStorage struct {
+	unimplementedPromStorageMethods
+	opts            Options
+	client          *http.Client
+	endpointMetrics map[string]instrument.MethodMetrics
+}
+
+func (p *promStorage) Write(ctx context.Context, query *storage.WriteQuery) error {
+	encoded, err := convertAndEncodeWriteQuery(query)
+	if err != nil {
+		return err
+	}
+
+	var wg sync.WaitGroup
+	multiErr := xerrors.NewMultiError()
+	var errLock sync.Mutex
+	for _, endpoint := range p.opts.endpoints {
+		endpoint := endpoint
+		if endpoint.resolution == query.Attributes().Resolution &&
+			endpoint.retention == query.Attributes().Retention {
```

Comment:
nit: As per effective go is better to use a continue if it doesn't match rather than using body inside the if case:
https://golang.org/doc/effective_go#if

> This is an example of a common situation where code must guard against a sequence of error conditions. The code reads well if the successful flow of control runs down the page, eliminating error cases as they arise. Since error cases tend to end in return statements, the resulting code needs no else statements.

So this could be:
```go
		if endpoint.resolution != query.Attributes().Resolution ||
			endpoint.retention == query.Attributes().Retention {
			continue
		}

		metrics := p.endpointMetrics[endpoint.name]
		wg.Add(1)
		go func() {
			defer wg.Done()
			err := p.writeSingle(ctx, metrics, endpoint.address, bytes.NewBuffer(encoded))
			if err == nil {
				return
			}
			errLock.Lock()
			multiErr = multiErr.Add(err)
			errLock.Unlock()
		}()
```